Refactor region codes 171

joseph-palmer opened the pull request:
Had a stab at changing how region codes are rendered.

This PR gives an additional field `region_codes` which defaults to `NULL` in `DataClass`. Each country now has a new method called `set_region_codes` which contains the contents of `region_codes.R` for that country, making a tibble from either data stored in a csv file in data-raw or, if there are not many codes, a tibble defined in code. 

When `get_region_codes` is now run it calls `set_region_codes` rather than the previos .rda file and then procedes as normal. The advantage of this is that now region codes are given as part of the class rather than having to add code to another file. 

### Issues

1. currently tests fail as the 'csv' files containing the region codes for some countried are set up as if the user is in the main 'covidregionaldata' working directory, so when ran from tests/testthat the paths are incorrect. This could be fixed by using a rda like the current region codes data set, except instead of having one we would have one for each country that needed one. However, would this would require authors to build and store these?  

2. Currently if region codes do not need to be set and `region_codes = NULL` the code errors with a custom message. It would be ideal to still render data without the region codes / handle this more elegantly. 

PR still needs some work but would like to get people's thoughts and sugestions, @seabbs @kathsherratt

RichardMN commented:
Just a note as I see this going forward. I like this new approach and it wasn't too bad to implement it for Belgium in #217 

I think that the possibility of saving these tables as CSVs (and in some cases, generating the CSVs and then importing them, as we do for France right now) is really helpful. CSVs are a lot more straightforward to edit (particularly for non-ASCII characters) and readable compared with a massive tribble/tibble definition which has to wrap to 80 character lines.

*Do* make clear (perhaps you already have) where the CSVs should be, whether/how they should be generated into an `rda`, and what other dependencies/expectations there would be on them.

I'll wait for this guidance and then I can probably apply it to my stable of PRs. (gaggle? flock?)

joseph-palmer commented:
Latest update:

Instead of a single `region_codes` rda file we now have seperate rda files for each country. This was done by placing the country specific parts of the code in `region_codes.R` into their own separate script, e.g. `render_italy_codes.R`. Only region codes which read from a csv file have their own script, for others such as Germany the code has been moved into a separate method of the country class. This method is called `set_region_codes` and stores the region code data from the rda file (or creates it in code) in a public field called region_codes. By doing it this way, shared methods to do with formatting the region codes for use can be inherited from `DataClass$get_region_codes()`. This function calls `set_region_codes` to store the country specific region codes in the public field, which can then be processed using the standard steps to create the lookup tables. 

Not all country classes have external region codes, e.g. `WHO`, `ECDC` and `UK` have region codes included in the dataset. In these cases a lookup table is not required but a code, which is used to name the column, is. In these instances, the country `set_region_codes()` return the code name as a string. This is then stored in a local variable in `get_region_codes` and set as codes if the public field `region_codes` is null, which it is by default.

joseph-palmer commented:
Also, this latest update should take care of #200

seabbs commented:
Just looking at this Joe, I am a bit confused why we need both `get_region_codes` and `set_region_code`?

seabbs commented on R/Canada.R:
```diff
@@ -24,10 +24,46 @@ Canada <- R6::R6Class("Canada",
+      )
+      canada_codes <- tibble::tibble(
+        country = "canada",
+        level = c("level_1_region"),
```
Why can't this just be stored as a field as its just data? Drop the country column as this is already built in the Canada class.

seabbs commented on R/Canada.R:
```diff
@@ -24,10 +24,46 @@ Canada <- R6::R6Class("Canada",
+        level = c("level_1_region"),
+        name = c("iso_3166_2"),
+        codes = list(
+          level_1_canada
```
only this needs to be assigned in this function and in fact maybe doesn't even need to be if it can be built from a field?

seabbs commented on R/ECDC.R:
```diff
@@ -28,6 +28,11 @@ ECDC <- R6::R6Class("ECDC",
     #' @field source_data_cols existing columns within the raw data
     source_data_cols = c("cases_new", "deaths_new"),
 
+    #' @description Specific function for getting country codes for ECDC .
```
just assign to a field surely?

seabbs commented on R/Germany.R:
```diff
@@ -28,6 +28,46 @@ Germany <- R6::R6Class("Germany",
+          "Niedersachsen", "Nordrhein-Westfalen", "Rheinland-Pfalz",
+          "Schleswig-Holstein", "Saarland", "Sachsen", "Sachsen-Anhalt",
+          "Th\u00FCringen"
+        )
```
same comments as in Canada

seabbs replied:
we only need a function to assign either level 2 or level 1 if we want to do that (or just assign both?)

seabbs commented on R/Italy.R:
```diff
@@ -25,6 +25,19 @@ Italy <- R6::R6Class("Italy",
+      message_verbose(
+        self$verbose,
+        paste(
+          "Getting region codes for",
```
same comments.

seabbs commented on R/UK.R:
```diff
@@ -169,7 +164,7 @@ UK <- R6::R6Class("UK", # rename to country name
         ) %>%
         # Join local authority codes to level 1 regions
-        left_join(self$authority_lookup_table,
+        left_join(covidregionaldata::uk_authority_lookup_table,
```
this doesn't need to have such a different structure to everything else does it?

seabbs commented on R/shared-methods.R:
```diff
@@ -102,28 +102,38 @@ DataClass <- R6::R6Class(
-        self$data$code <- codes$name[[1]]
-        self$data$codes_lookup <- codes$codes[[1]]
+      code_name <- self$set_region_codes()
+      if (is.null(self$region_codes)) {
```
I don't think we need this function?

RichardMN replied:
I was surprised when I went looking (trying to figure out something else) the range of things which `get_region_codes` does. It seems very much an initializer (setting up `self$data` is pretty key). If it were given a different name and the region codes functionality were broken out somewhere else would this work better? (I tried calling `get_region_codes` twice with unfortunate consequences.)

seabbs commented on data-raw/render_italy_codes.R:
```diff
@@ -0,0 +1,13 @@
+# italy codes
+italy_codes <- vroom::vroom("data-raw/italy_codes.csv")
+italy_codes <- tibble::tibble(
```
all looks sensible but I don't think the field info is needed here (as it should go in the class)

seabbs reviewed (changes requested):
Some general thoughts. Think might need a bit of a redesign to be simpler.

joseph-palmer commented:
So my thinking is `set_region_codes` will just pull in the region codes (either from a file or create them on the fly for some) but the turning this into a lookup table and codes etc. is done for all countries, so to save having to type this out in every country as `get_regional_data` I set it in the parent class and just have `set_regional_data` pull out the region codes and hand these back to `get_region_codes`. 

I may be over thinking this however....

joseph-palmer commented:
Ahhhh I see. Sorry should have spotted this sooner, you're right `set_region_codes` isn't needed it can just be assigned to the class public field. I will update this.

seabbs commented:
and that `get_region_codes` isn't needed. The only reason we might want `set_region_codes` is to change fields depending on the level of the data required which is something we might want. 

So this option would be something like:

```
set_region_codes <- function() {
if (self$level == 1) {
   self$region_code =  self$level_1_region_code
   self$codes = self$level_1_codes
}else{
  lself$region_code = self$level_2_region_code
  self$codes = self$level_2_codes
}
}

joseph-palmer commented:
but `get_region_codes` sets them for use in clean and processing. How would this work without `get_region_codes`?

seabbs commented:
it only does that because that is how its coded. It doesn't need to work like that?

joseph-palmer commented:
Sorry, I'm having trouble visualising this.

The region codes are stored in a nested tibble where 'level_?_region' contains a dataset of two columns which are the region name and the code for that region. The code in `get_region_codes` uses the level requested (`self$level`) to get the level of the tibble, pull out the required region codes for that level and then sets this as a field so it can be joined on to the data in the cleaning step. The codes are also used in the processing step as column headings. As it stands each country class needs to provide the region codes and they are handled in the same way by `DataClass`. 

If we don't have `get_region_codes` then this would need to be included in the cleaning step of every country?

joseph-palmer commented:
@seabbs I've just done this on Italy so far just to get your thoughts on if this is looking ok and see if there is anything mising.

I have removed `get_region_codes` in `DataClass` and now have a `set_region_codes` which pulls the data into a named list based on the column 'level' (e.g. `region_codes = list("level_1_region" = data_for_level_1, etc)` ). This runs at initalization. This set up should allow country specific things, such as nhs codes, to be added easily.

I have also updated `process_internal` to work with passed arguments. 

This seems to work OK for Italy but just want to check its on target before extending to the other countries (at the moment they will not work with these changes). 

One thing I will need to work out is how to format the currently nested tibbles for the others as single dataframes. This was easy for Italy as there is only one level, but not sure for the others exactly how to acheive this.

seabbs commented:
I'll take a look now @joseph-palmer  all sounds good. 

To reformat something like this (level could be linked to the meaningful level name for Mexico and that might be the better choice rather than "1").

```
Mexico_codes <- level_1_codes %>%
   mutate(level = "1" ) %>%
   rename(region = level_1_region, code = level_1_region_code) %>%
   bind_rows(
    level_1_codes %>%
       mutate(level = "2" ) %>%
       rename(region = level_2_region, code = level_2_region_code) 
  )
```

All from `dplyr`.

seabbs reviewed (commented):
All looks good to me.

seabbs commented on R/processing.R:
```diff
@@ -186,12 +186,13 @@ totalise_data <- function(data) {
-process_internal <- function(region, group_vars, totals = FALSE,
-                             localise = TRUE, verbose = TRUE) {
-  if (!any(class(region$clean) %in% "data.frame")) {
+process_internal <- function(clean_data, level, code, group_vars,
```
looks good. Need to update documentation in line with this change and potentially testing (if `process_internal` is being tested outside of `process` which maybe it isn't.

seabbs replied:
need to document level and code

seabbs commented on R/shared-methods.R:
```diff
@@ -72,7 +72,7 @@ general_init <- function(self, level = "1",
   self$steps <- steps
   self$country <- tolower(class(self)[1])
-  self$get_region_codes()
+  self$set_region_codes()
```
add default method (likely just empty)?

seabbs commented on R/shared-methods.R:
```diff
@@ -102,29 +102,14 @@ DataClass <- R6::R6Class(
     steps = NULL,
+    #' @field region_codes string or tibble Region codes for the target country
+    region_codes = list(),
+    #' @field code_name
```
how does this deal with level 1 and two differences? Where are we setting this?

seabbs replied:
presumably needs to be set in the Italy class.

seabbs commented on R/shared-methods.R:
```diff
@@ -102,29 +102,14 @@ DataClass <- R6::R6Class(
-      }
+    #' @description Place holder for custom country specific function to load
+    #' region codes.
+    set_region_codes = function() {
```
ignore previous comment your ahead of me 😜

seabbs commented on R/shared-methods.R:
```diff
@@ -151,8 +136,9 @@ DataClass <- R6::R6Class(
       )
-      self$data <- process_internal(
-        self$data,
+      tar_level <- paste0("level_", self$level, "_region")
```
LGTM

seabbs commented on data-raw/render_italy_codes.R:
```diff
@@ -0,0 +1,10 @@
+# italy codes
+italy_codes <- vroom::vroom("data-raw/italy_codes.csv")
+italy_codes <- tibble::tibble(
+  level = c("level_1_region"),
```
obviously needs updating to new approach.

seabbs replied:
see main thread comment for some pseudo code.

seabbs reviewed (commented):
First pass missed quite a few things. Looks good on the second pass as well.

seabbs commented:
Apologies all for the lack of work on this, should be back on it tomorrow and hopefully get this all somewhere stable.

seabbs reviewed (dismissed):
I think this is now all in place. @joseph-palmer please review as well even though it can't be formally as you opened the PR. 

This PR now: 

- Adds supported levels, region names, and region codes as lists to the class object 
- Adds `set_region_codes` which loads `codes_lookup` as a list either from data in code or from a package dataset. 
- Updates processing code to be robust to missing columns
- Streamlines and cleans up some elements of the implemented region code
- Extends `check_country_` to handle some types of fuzzy name matching
- Adds a `summary` method to all classes which is now used by `get_availablea_datasets` to assign a table 
- Adds a `get` method that combines all other methods. Adds a `get` arg to `new` so that the following can be used to return a dataset
```
g <- Germany$new(get = TRUE)
g$return()
```
- Updates tests to handle new changes
- Streamlines `get_regional_data` to use the `get` method. 
- All of above implemented in all classes currently supported on `dev`.

seabbs reviewed (approved):
Super amazing work as always @seabbs! (I am taking the silence as assent and merging - we can fix once on dev but PRs are getting a bit backed up).

kathsherratt commented:
Thanks! Sorry, I just haven't had time to review yet - but seems likely I'd agree with your review anyway... and agree good to merge into dev and make some progress!
